sdk/iot/core/src/az_iot_hub_client.c

```diff
@@ -1,12 +1,137 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // SPDX-License-Identifier: MIT
 
+#include <stdint.h>
+
 #include "az_iot_hub_client.h"
+#include <az_precondition_internal.h>
+#include <az_result.h>
 #include <az_span.h>
 
 #include <_az_cfg.h>
 
+static const uint8_t hub_client_forward_slash = '/';
+static const uint8_t hub_client_param_separator = '&';
+static const uint8_t hub_client_null_terminate = '\0';
+
+static const az_span hub_client_api_version = AZ_SPAN_LITERAL_FROM_STR("?api-version=2018-06-30");
+
 AZ_NODISCARD az_iot_hub_client_options az_iot_hub_client_options_default()
 {
   return (az_iot_hub_client_options){ .module_id = AZ_SPAN_NULL, .user_agent = AZ_SPAN_NULL };
 }
+
+AZ_NODISCARD az_result az_iot_hub_client_init(
+    az_iot_hub_client* client,
+    az_span iot_hub_hostname,
+    az_span device_id,
+    az_iot_hub_client_options const* options)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(iot_hub_hostname, 1, false);
+  AZ_PRECONDITION_VALID_SPAN(device_id, 1, false);
+  AZ_PRECONDITION_NOT_NULL(options);
+
+  client->_internal.iot_hub_hostname = iot_hub_hostname;
+  client->_internal.device_id = device_id;
+  memcpy(&client->_internal.options, options, sizeof(az_iot_hub_client_options));
```

Comment:
If we are not going to add more members to `az_iot_hub_client_options`, we can expand `memcpy` to two assignments for both code size and performance.

Reply:
This is not guaranteed to stay at two options but for now that's what we have. I can revert back for now if assignments are better than `memcpy`